### ISSUE TITLE
x/tools/internal/lsp: does not handle build tags

### ISSUE THREAD

**Issue by stamblerre** [golang/go#29202]
We need to handle build tags correctly in the LSP.

**Comment by glerchundi** [comment]
Bitten by this also, lost an hour trying to figure out why after migrating to `gopls` basic functionalities didn't work...no import-organize, no format, no nothing. Afterwards I found that it's was working properly for everything except for files with build tags as mentioned in #31286.

**Comment by myitcv** [comment]
I'm wondering whether this is the right issue to become the umbrella issue for our long-running discussion about context?

**Comment by sjdweb** [comment]
This is killing me with integration tests that have build tags. Any workarounds known?

**Comment by christiankiely** [comment]
@sjdweb current workaround I'm using is to set the GOFLAGS environment variable. This is the configuration I'm using in VS Code:

```
"gopls.env": {
        "GOFLAGS": "-tags=test"
    },
```

**Comment by myitcv** [comment]
This was brought up on Tuesday's [golang-tools](https://github.com/golang/go/wiki/golang-tools) call. 

@christiankiely - the "editor level" `GOFLAGS` approach your describe works in some cases, but not all. Because setting the flag in that way sets the flag for _all_ modules and packages, i.e. for the `gopls` instance used by the editor.

In discussions we've had thus far it seems for a complete solution we need to be able to define the sets of valid build flags at various levels:

* Editor level (i.e. throughout editor)
* Module level
* Package (tree) level
* File level

We also need:

* the ability to specify, from within the editor, which build flags are in operation, because for a given file/package/module there might be multiple valid options
* the ability to then pass this choice on a per request basis to `gopls`

The conclusion on Tuesday's call was broadly that there are other higher priority issues right now; add to that, a solution that doesn't involve LSP spec changes isn't exactly jumping out at us.

cc @ianthehat @dominikh

**Comment by breml** [comment]
Bitten by this as well. I use build tags to distinguish between normal builds and debug builds. Now VSCode reports `logError redeclared in this block` error and the file is constantly marked as having errors.

**Comment by gracenoah** [comment]
Another populate use case: when using [wire](https://github.com/google/wire), there are two copies of the initializers:
1. The template in a file with a build tag that never gets compiled
2. The real one generated by wire with no build tags

gopls highlights them as duplicate definitions because it ignores build tags.

Is there a short term workaround in this simpler use case? Some way to make gopls ignore files with the wireinject build tag?

**Comment by stamblerre** [comment]
Thank you for sharing this info. We realize that this is an issue that affects a lot of people, so we're trying to come up with a viable solution. It's a complicated problem, but to start, I think we can try to disable diagnostics on files that would be excluded given your GOOS/GOARCH.

**Comment by evanphx** [comment]
As a data point, I'm experiencing this today in any package that picks up `os/user` because it uses build tags to decide how to wire in OS user lookup. In my case, it's that analysis fails because `cgoLookupHost` isn't defined as files aren't being evaluated.

I understand that you'd like for gopls to allow for reading and analysis of files intended for a different GOOS that the one that the editor is on, but given how packages are structured, you'd having to perform analysis on the matrix of build tags. That seems.... time consuming at best. Plumbing through a default tag set seems like an easy way to at least get folks moving again.

I'm happy to look into how to do that if need be! Please just let me know.

**Comment by evanphx** [comment]
Looks like it's also firing from `net` looking like:

`/usr/local/Cellar/go/1.12/libexec/src/net/lookup_unix.go:81:24: undeclared name: cgoLookupHost`

**Comment by zchee** [comment]
@evanphx I also found this issue.
The workaround is launch `gopls` use `CGO_ENABLED=0` or something. Note that might be vscode can't that workaround AFAIK. Also not solved if set `CGO_ENABLED=0` to `env`. Because it is late to place.
https://github.com/golang/go/issues/29202#issuecomment-488469829

Now I develop the `client` side for IDE written in Go, it's implemented set some `env` to launch `gopls` binary using set to `exec.Command.Env`. I did solve the use it.
Anyway, We should including `CGO_ENABLED=0` to `gopls` send `Build info`.

**Comment by evanphx** [comment]
@zchee Ah! That makes a lot of sense, I'll start it with CGO_ENABLED=0 and see how it works.

**Comment by zchee** [comment]
@evanphx 
Also, I was send CL for support build tags.
https://go-review.googlesource.com/c/tools/+/178782

**Comment by zchee** [comment]
@evanphx
FYI, ref for your issue root cause:
https://github.com/golang/go/issues/31705

**Comment by zchee** [comment]
@stamblerre 
Can we close this issue by below CL?

https://github.com/golang/tools/commit/70bf279967a6c40a5e75fb4a31ec8273089700f8

**Comment by stamblerre** [comment]
I think we will still need special handling for build tags so that users don't need special configurations to get `gopls` features in files tagged for different build systems.

**Comment by zchee** [comment]
@stamblerre 
I see, totally agreed.

**Comment by davidchen0310** [comment]
> @sjdweb current workaround I'm using is to set the GOFLAGS environment variable. This is the configuration I'm using in VS Code:
> 
> ```
> "gopls.env": {
>         "GOFLAGS": "-tags=test"
>     },
> ```

This incurs an `Unknown configuration setting` error in `VS Code` 1.36.1 with `gopls` v0.1.3, and instead I have to use the configuration below to make it work.

```
"go.toolsEnvVars": {
    "GOFLAGS": "-tags=functional"
},
```

**Comment by stamblerre** [comment]
I recently opened https://github.com/golang/go/issues/33389 to get a better sense of the use cases for build tags. Please add your own use cases if they are not yet listed in this issue - it will help us better understand how to solve this issue.

**Comment by stanleynguyen** [comment]
+1 
I'm having issues with VSCode flagging my different function implementations for different versions of go as duplications

**Comment by KBassford** [comment]
RS: You probably already know this, gopls does not handle mutually exclusive build tags at all.  I have one file with '// +build debug', and an almost identical file in the same folder with '// +build !debug'.  gopls flagged every constant, variable and function as being redefined between the two files.  Good luck with the fix.

**Comment by mvdan** [comment]
https://github.com/golang/go/issues/39005 might be of interest to the people following this thread. It's an attempt at defining a build-system-agnostic list of build configurations for a project, in a way that can be consumed by many tools like staticcheck and gopls.

**Comment by lucaswxp** [comment]
For anyone having issues with wire + VSCode flagging auto generated function implementations as duplications, edit the project settings json and put this:

```json
"go.buildTags": "wireinject"
```

This solved for me. Have a downside that now your wire.go file will not have any code completion and validation, but you  can just uncomment the build flag when developing to get those back.

**Comment by jackielii** [comment]
in coc.nvim+coc-go, for wire inject, add:
```
  "go.goplsOptions": {
    "buildFlags": ["-tags=wireinject"]
  },
```

**Comment by cyberbeast** [comment]
Is there any timeline on releasing support for build tags in gopls, and consequently in vscode?

**Comment by stamblerre** [comment]
We will not add any new behavior to improve build tag handling before `gopls/v1.0`. For now, the best way to work with `gopls` and build tags is by using the configurations suggested above.

**Comment by cyberbeast** [comment]
@stamblerre What's the exact property that needs to be set in settings.json (vscode) to switch build tags?

**Comment by stamblerre** [comment]
I would suggest the work-around described in this comment: https://github.com/golang/go/issues/29202#issuecomment-488469829.

**Comment by ericwq** [comment]
for a variable being overwritten based on build constraints,  I tried every suggested config in this thread. it doesn't work.

here is the overwritten case:

- when edit a file, such as [internal/lsp/lsprpc/lsprpc.go](https://github.com/golang/tools/blob/master/internal/lsp/lsprpc/lsprpc.go) , 
- you see a  statement in line 289, "func _connectToRemote_":
- "network, address = _autoNetworkAddress_(goplsPath, inAddr)",
- now, to find the _autoNetworkAddress_ func, you send a 'definition' query to the gopls server.
- in current implementation,  'definition' query will return the [internal/lsp/lsprpc/autostart_default.go](https://github.com/golang/tools/blob/master/internal/lsp/lsprpc/autostart_default.go) result. and the editor jump to that file directly. this implementation will mislead the developer. i checked the code several times and finally notice the 'build tags'

the propose is to return the [internal/lsp/lsprpc/autostart_posix.go](https://github.com/golang/tools/blob/master/internal/lsp/lsprpc/autostart_posix.go) , the behavior of 'definition' query should consider the 'build tags', in this case, it is "// +build darwin dragonfly freebsd linux netbsd openbsd solaris".

any suggestion on this case? @stamblerre

**Comment by sheldoncoupeheure** [comment]
To be able to edit windows-specific golang code ("xxx_windows.go") requiring windows-only golang packages on my OSX machine, I had to add this to my VSCode settings:

```js
"go.toolsEnvVars": {
    "GOOS": "windows"
 }
```

**Comment by Merrit** [comment]
```
"gopls.env": {
        "GOFLAGS": "-tags=test"
    },
```

Where does this go? I tried settings.json and I get:
> Unknown Configuration Setting

**Comment by stamblerre** [comment]
@Merrit: That's correct--the unknown configuration setting message is a warning, but it will still work.

**Comment by Merrit** [comment]
I see. Am I misunderstanding the purpose then?

For example in one of my _linux.go files I am still getting `No packages found for open file`, unable to access functions from this file, etc.

If I add `// +build windows` everything works as expected.

**Comment by stamblerre** [comment]
Are you working on Windows / adding the Windows tag to the Linux file? OS-specific tags are detected based on your OS, so unless you set `GOFLAGS` to `-tags=linux`, you won't get editor features working for the _linux.go files.

**Comment by Merrit** [comment]
Yes, currently working on Windows & having issues working with *_linux.go files. Not sure why I would _want_ to be unable to reference or work on Linux files just because I am on Windows, or vice versa.

Setting this in settings.json seems to have it working, so far:

```json
  "gopls.env": {
    "GOFLAGS": "-tags=linux"
  },
```

Thank you!

**Comment by Merrit** [comment]
Spoke too soon.

Editor features work fine, but it then fails to run / debug.

`undefined: funcFromLinuxFile`

I tried adding `"buildFlags": "-tags=linux"` to launch.json, but that makes it try to compile the linux version which fails quite spectacularly. :)

Also tried `"buildFlags": "-tags=test"`, `"buildFlags": "-tags=!linux"` and `"buildFlags": "-tags '!linux test'"`.

Only solution so far is to just stop naming things with the *_os.go scheme, which I was given to understand was the preference for conditional compilation. FYI, naming the files `featureLinux.go` and leaving it be works just fine. Adding `// +build linux` to that same file breaks things the same as using the *_linux.go naming scheme.

**Comment by stamblerre** [comment]
> Not sure why I would _want_ to be unable to reference or work on Linux files just because I am on Windows, or vice versa.

This is a current limitation of `gopls`, and this issue is tracking improving that behavior.



> I tried adding `"buildFlags": "-tags=linux"` to launch.json, but that makes it try to compile the linux version which fails quite spectacularly. :)

You won't actually be able to run code meant for Linux on Windows--my suggestion was just to get editor features for development purposes.

**Comment by Merrit** [comment]
Right, I don't want to run or compile anything meant for a different OS. Just mentioning what I had tried to resolve my issue.

Ideally what I would like is:
1. Name files by OS to take advantage of conditional compilation, ie feature_windows.go & feature_linux.go
2. Editor features still work, eg it recognizes the packages, doesn't complain about undeclared variables or functions, etc just because those variables/functions are in a feature_linux.go when I am on Windows, or vice versa.
3. Still be able to use VSCode's run/debug features (for the code related to the OS I am on).

Currently I can only get 2 & 3 on my list by _not_ naming files by OS nor using the `// +build _____` -- which I understand means no conditional compilation.

**Comment by FrankReh** [comment]
Wanted to be able to use _go generate_ in a directory to get a _go run gen.go_ called and found gopls doesn't seem to work on the gen.go file because of a mix of build flags in other local files. Can I work around this by invoking vim on gen.go with an environment variable set to something for the benefit of gopls or is moving the gen.go into its own directory my best option?

_gopls definition_ does work from a file that has an _ignore_ build tag, but not if a plain go file in the same directory exists.

```sh
$ cat -n gen.go 
     1	// +build ignore
     2	
     3	package main
     4	
     5	import (
     6		"fmt"
     7	)
     8	
     9	func main() {
    10		fmt.Println("Hello world!")
    11	}
$ gopls definition gen.go:10:10
gopls: /<snip>/gen.go:10:10: no packages returned: packages.Load error
$ rm config.go
$ gopls definition gen.go:10:10
/<snip>/go1.15.5/src/fmt/print.go:273:6-13: defined here as func fmt.Println<snip>
 ```

**Comment by stamblerre** [comment]
The command line doesn't really give you much configurability, but you can set the `GOFLAGS` environment variable before running the `gopls` command. Otherwise, you can configure `gopls.buildFlags` or `GOFLAGS` as described in the examples above.

**Comment by VikingPingvin** [comment]
I am facing this issue as well, appending `settings.json` with ```"gopls.env": {
            "GOFLAGS": "-tags=debug"
          },```does not resolve the issue for me.

In my case I have two files `thermal.go` and `thermal_mock.go`, with build tags `// +build !debug` and `// +build debug` respectively, so I can mock target HW specific functions.

My biggest issue is that Intellisense completely stopped working in the `_mock` version, and the gopls.env in `settings.json` didn't help either. 

Go 1.15.5, gopls v0.6.0

Any follow up on the issue?

**Comment by stamblerre** [comment]
> In my case I have two files `thermal.go` and `thermal_mock.go`, with build tags `// +build !debug` and `// +build debug` respectively, so I can mock target HW specific functions.
> 
> My biggest issue is that Intellisense completely stopped working in the `_mock` version (as in, worked the other day, next day poof) , and the gopls.env in `settings.json` didn't help either.

Would be happy to help investigate this, but please file a new issue with the following information: https://github.com/golang/tools/blob/master/gopls/doc/troubleshooting.md.

**Comment by dylankb** [comment]
Note for VSCode settings.json in case it's helpful. Before upgrading to 0.6.2 I used to use `GOFLAGS`.

```
"gopls": {
  "env": {
    "GOFLAGS": "-tags=<tag1>,<tag2>"
  },
},
```

After upgrading I started getting `No packages found for open file...` errors. I switched `go.buildFlags` and no longer use `GOFLAGS` without issue it seems.

```
go.buildFlags: "<tag1>,<tag2>",
```

**Comment by kevlar700** [comment]
The fixes here didn't work for me.

I found 
env GOOS="openbsd" code

Worked for me at the editor level

P.s. Irrelevant tip: if you also set HISTFILE and open a project folder after "code" then you can have dedicated terminal history and a project folder opened together.

**Comment by serezlan** [comment]
Hello,

I'm facing this issue on linux using emacs. Have tried suggestions above and the one that is working is to delete and re-adding build tag line.
I also tried using VS Code with the same result.

**Comment by jalseth** [comment]
I was finally able to get this working in VS Code by using the following and restarting VS Code.

```json
{
    "go.buildTags": "tag1,tag2"
}
```

**Comment by laeo** [comment]
I think gopls shouldn't care about build constraints, or it only cares about the file I'm editing, which I call "constraint aware".

**Comment by calebstewart** [comment]
This and #49657 were the only relevant results for my internet searches, so, for historical reasons, my solution for `emacs` is to use `setenv` at the global scope and modify `GOFLAGS`. This isn't perfect but has worked for me:

```elisp
(setenv "GOFLAGS" "-tags=linux,windows")
```

Is there a plan to make something like this a permanent in `gopls`? Is there some problem with just specifying multiple platform tags by default? It hasn't seemed to cause any problems for me specifically. It doesn't make much sense to me that `gopls` would care about the platform, and simply failing for platform-tagged files breaks the entire workflow of developing cross-platform applications. This is almost certainly an undocumented bug in `gopls`, unless I missed some documentation some where (which I'd be happy if someone shared :smile:).

**Comment by hyangah** [comment]
@calebstewart That's insteresting. `gopls` heavily depends on `go list` and other `go` commands to parse go code, and by your setting `-tags=linux,windows`, the flag is passed to all the go commands `gopls` uses. If you see `gopls` and so `go` commands work with `-tags=linux,windows`, that's by accident, not intentional. 

And, more puzzling thing is that AFAIK platform (os/arch) are special and go commands select them based on `GOOS`/`GOARCH` environment variables, not based on the `-tags` flag in most cases. So, I am not sure what's working and what's not working with specifying OSes using `-tags`.

**Comment by hsiaosiyuan0** [comment]
for vscode users, update your `PROJECT_ROOT/.vscode/settings.json` like below my resolve this problem:

```json
{
  "go.toolsEnvVars": {
    "GOOS": "js",
    "GOARCH": "wasm"
  }
}
```

**Comment by calebstewart** [comment]
@hyangah interesting, it could have been some configuration specific to this project (it's a work project, and I'm not the one who configured it), then, but that seemed to be the best approximation of the suggestions I was seeing online. Admittedly, I'm not exactly sure why it worked for me either, but all I can say is that it did. :eyes:

**Comment by ahmafi** [comment]
I have a similar problem, I have separate files (`source_windows.go`, `linux_windows.go`, `darwin_windows.go`) for setting platform-specific configs, and by default, I don't get any code completion on Windows or darwin codes (I'm using VSCode on Linux) and it shows the following message:
```
No packages found for open file /home/amir/projects/Memor/memor/pkg/config/config_darwin.go: <nil>.
If this file contains build tags, try adding "-tags=<build tag>" to your gopls "buildFlags" configuration (see (https://github.com/golang/tools/blob/master/gopls/doc/settings.md#buildflags-string).
Otherwise, see the troubleshooting guidelines for help investigating (https://github.com/golang/tools/blob/master/gopls/doc/troubleshooting.md).go list
```
after adding the following tags I'll get a duplicate declaration error on these files because I'm declaring the same things in these 3 files.

``` json
"gopls": {
  "build.buildFlags": ["-tags=linux,windows,darwin"],
}
```

**Comment by findleyr** [comment]
@Michael-F-Ellis you should be able to work on xmain OR ymain by setting just one of those two build tags, the same as you do when you run `go build -tags xmain`. As described in https://github.com/golang/go/issues/49104#issuecomment-1008314902, `go vet` and `go build` have the same limitations as gopls: only one set of build tags can be active at a given time.

The critical missing feature for gopls is the ability to edit using multiple sets of build tags simultaneously.

**Comment by syncjuncture** [comment]
Bump

**Comment by alexgille** [comment]
Hello,

Is this topic still in progress or should we not expect any improvement regarding this limitation?

Thanks,
Best regards

**Comment by findleyr** [comment]
Hi all, this issue deserves an update (and sorry @alexgille for missing your inquiry above). Let me summarize the current state of things:

We know that support for multiple sets of build tags is one of the most wanted features. Unfortunately, build tags pose a fundamental problem that we can't fix without changing many other things about gopls. Because many of gopls' algorithms rely on having an active graph of `go/types.Packages` in memory, and `go/types` relies on the canonical identity of its objects, gopls can't simply process multiple sets of build tags without maintaining two copies of the package graph (because ~everything will at some point in its transitive imports depend on a package for which there are multiple definitions based on build tags). The package graph drives most of gopls' memory usage, so maintaining N copies of the package graph corresponds to approximately an N-fold increase in memory and performance. Given that memory usage is a huge pain point for our users, a solution that depends on throwing memory at the problem will probably not be viable for many users.

Even if that solution is acceptable for some users with small workspaces, it poses a significant burden on the code in an area where we have historically had a lot of bugs: package metadata tracking and package invalidation. I am currently trying to refactor that area of gopls and fix all the bugs. In gopls@v0.9.0 the representation of metadata was refactored, which contributed to the [performance improvements](https://github.com/golang/tools/releases/tag/gopls%2Fv0.9.0) observed in that release. In gopls@v0.9.2, I'm trying to [fix all the outstanding bugs we know about](https://github.com/golang/go/issues?q=is%3Aissue+label%3Agopls%2Fmetadata+milestone%3Agopls%2Fv0.9.2) related to metadata. So here also, we are hesitant to throw more complexity at the problem.

Longer term, an even higher priority for us is to make gopls scale better and load faster. In order to achieve this, we will inevitably need to break our dependence on the package graph. The first step toward this is rewriting key algorithms to be based on data structures that can be incrementally updated at the package level. Once we do this, the problem of handling multiple sets of build tags becomes tractable.

So resolving this issue can't realistically happen soon, but is part of our longer terms goals for gopls.

In the meantime, I think we can still improve the status-quo. For example:
 - we can make it easier to switch the set of active build tags (through better handling of configuration changes, and code lenses on build constraints, for example)
 - we can resolve #49657, making it easier to work with packages defined by a single build tag (e.g. `ignore` or `mage`).

**Comment by Torwalt** [comment]
For nvim-lspconfig users (with nvim-lsp-installer) following settings worked for me:

```lua
lspconfig.gopls.setup {
    settings = {
        gopls = {
            env = {
                GOFLAGS = "-tags=windows,linux,unittest"
            }
        }
    }
}
```

with files looking like

thing_test.go:

```go
//go:build unittest

package thing_test
```

and thing.go

```go
package thing
```

all lsp features work.

**Comment by eric-burel** [comment]
Cross-posting from https://github.com/golang/vscode-go/issues/1799:

Is there an alternative without relying on setting env variable in VS code settings? 
VS code doesn't support nesting settings so it's annoying when you mix go server code and go wasm code or more broadly different architecture.

Maybe a top-level comment in the file? This seems to have been mentioned at the begging of this thread: https://github.com/golang/go/issues/29202#issuecomment-488590710, https://github.com/golang/go/issues/29202#issuecomment-498926547

(there are many messages, sorry if it has already been mentioned/answered)

**Comment by inliquid** [comment]
@eric-burel as a workaround you can open projects in different VS Code instances. WASM client code can be placed somewhere in project dir and can have it's own workspace settings in `.vscode/settings.json`. I'm using it this way for many years already.

**Comment by 404cn** [comment]
For emacs and eglot users:

```elisp
(setq-default eglot-workspace-configuration
              '((:gopls
                 (:ui.completion.usePlaceholders . t)
                 (:ui.diagnostic.staticcheck . t)
                 (:build.buildFlags . ["-tags=wireinject"]))))
```

**Comment by Brandon-M-Hartman** [comment]
Has there been any update on this?  Trying to build wasm plugins for my go program and this is obnoxious to deal with.

**Comment by zzxn** [comment]
This problem is still there, att least a more friendly diagnose tip should be given.

**Comment by arnevm123** [comment]
@Torwalt 
This still gives a `compiler: TestFunc redeclared in this block (see details) [DuplicateDecl]` , but I guess it's an improvement :smile: 
Also I can not get this to work with //go:build windows in one file and //go:build !windows in another. (I guess this makes sense because these can't both be true at the same time...)

You can also use this instead of the env, which I find a bit cleaner:

```
settings = {
    gopls = {
        buildFlags = { "-tags=!windows,windows,darwin" },
    },
},
```

**Comment by mxygem** [comment]
I currently work in a repo that uses build tags for test files VS Code throws following (inaccurate) error about workspaces when a build flag is present. Setting the tags in the env variables sent to gopls does work around it, I just wanted to inform of another instance of this issue.

```
This file is within module ".", which is not included in your workspace.
To fix this problem, you can add a go.work file that uses this directory.
See the documentation for more information on setting up your workspace:
https://github.com/golang/tools/blob/master/gopls/doc/workspace.md.go list
```

**Comment by findleyr** [comment]
@mxygem can you say more about how you're using build tags in test files? Which build tags, and how are you using them?

**Comment by solidiquis** [comment]
Omg @Torwalt thank you!!

**Comment by findleyr** [comment]
Hi folks! I'm working on landing improvements to build tag support for gopls@v0.15.0.  Specifically, https://go.dev/cl/551897 added logic so that when you open a file constrained by GOOS or GOARCH, gopls will try to find a GOOS/GOARCH combination that includes that file, and will start tracking a build for the new configuration.  This means that most gopls functionality will Just Work when, for example, you open up a `*_darwin.go` file while working on linux.

If you're interested in being an early tester, you can install the first prerelease:

```
go install golang.org/x/tools/gopls@v0.15.0-pre.1
```

In the workaround suggested above (https://github.com/golang/go/issues/29202#issuecomment-1233042513), gopls loads only one build configuration, modified to include files for both linux and windows, but this will often lead to spurious build errors for duplicate declarations.  Now that it supports multiplexing across multiple builds (#57979), gopls can finally support multiple GOOS/GOARCH combinations simultaneously. If you are testing out the prerelease, please remove any `buildFlags` configuration from your settings.

If you encounter problems, please let me know and revert to gopls@latest. This is a rather large change to the way gopls operates, and so will require some time to polish before it can be released. Among the potential new bugs, we may encounter new performance edge cases, or possibly conflicting diagnostics from different builds (though I have tried to mitigate both). Additionally, we may require additional UX improvements to make it clearer which build configuration is being used for a given file.

One other area where we could improve build tag support is by automatically handling go version constraints (e.g. `//go:build !go1.20`). After consulting with the team and Go command experts such as @bcmills and @matloob, we think it is too complex and potentially problematic to support dynamic builds at different Go versions. We also expect that this is an order of magnitude less important than GOOS/GOARCH support -- let us know if you disagree.

**Comment by meling** [comment]
@findleyr re:
> @mxygem can you say more about how you're using build tags in test files? Which build tags, and how are you using them?

I may have a similar scenario to what @mxygem described [here](https://github.com/golang/go/issues/29202#issuecomment-1804652380).

I have a number of exercises for students, where I provide a template to ensure that our tests are compatible with a student submitted solution. Here is an example:

Template provided to students (`rot13.go`):
```go
//go:build !solution   // This line is actually removed before we provide it to students.

package cipher

type rot13Reader struct{ r io.Reader }

func (r rot13Reader) Read(p []byte) (n int, err error) {
	return 0, nil
}
```

Test provided to students (`rot13_test.go`):
```go
func TestRot13(t *testing.T) { ... }
```

We also have a solution file (`rot13_sol.go`) that we don't provide to students:
```go
//go:build solution

package cipher
```

These files are in the same folder like this:
```console
cipher/
├── rot13.go
├── rot13_sol.go
└── rot13_test.go
```

From the command line I can test that the template version fails because the `solution` tag is not set:
```console
% go test
--- FAIL: TestRot13 (0.00s)
    rot13_test.go:33: rot13.Read("Go programming is fun."): (-want +got):
...
```

And similarly, that the `rot13_sol.go` version passes the tests because the `solution` tag is set:
```console
% go test -tags solution
PASS
ok  	dat520/lab1/gointro/cipher	0.280s
```

This approach is convenient because I can keep the template, test, and the solution in same folder.

With VSCode, I can do:
```json
    "gopls": {
        "buildFlags": [
            "-tags=solution"
        ]
    },
```
To enable compilation of the `rot13_sol.go` file, and switch to `-tags=!solution` to enable compilation of the `rot13.go` file.

However, I'm unaware of any solution to avoid the error:
```text
No packages found for open file ... /lab1/gointro/cipher/rot13.go.
This file may be excluded due to its build tags; try adding "-tags=<build tag>" to your gopls "buildFlags" configuration
See the documentation for more information on working with build tags:
https://github.com/golang/tools/blob/master/gopls/doc/settings.md#buildflags-string.
```

Apologies for the long message, and the fact that I didn't take the time to read through the entire issue history. Anyway, I hope this can serve as another example of how tags are being used; I suspect others use it similarly.

**Comment by findleyr** [comment]
@meling thanks for explaining your use-case. Indeed, there's no way to avoid the error you report in whichever file is excluded by the current configured build tags. In general, there's nothing we can do about this _automatically_, since it's not possible to guess the set(s) of build tags to use.

Theoretically, since we now support multiple builds per folder, we could accept something like

```
    "gopls": {
        "buildFlags": [
            [],
            ["-tags=solution"]
        ]
    },
 ```
 
I.e., overload the "buildFlags" setting to accept multiple sets of flags.  I'd really like to avoid this if possible, because it complicates the configuration significantly. I wonder how many others would be interested in configurations like this.

**Comment by inliquid** [comment]
@findleyr is there any point in using "gopls.BuildFlags" instead of "go.buildTags"?

**Comment by findleyr** [comment]
@inliquid I believe "gopls.buildFlags" affects only gopls, whereas "build.buildFlags" affects all functionality (gopls, test, linter, debugger, etc.), which may or may not be desirable.  "go.buildTags" is an older configuration mechanism that was generalized to "buildFlags".

**Comment by meling** [comment]
@findleyr thanks for your reply. 

For my "simple" use-case where there is a binary choice, would it be possible to recognize the tag of the current editor window and just remove the error from source files whose tag is the negation? And then switch if the active editor window changes to the negated tag.

It is not clear to me if dynamically "guessing" based on active editor window is possible.

**Comment by findleyr** [comment]
@meling you're right, we could do that if there is only one build tag, though probably not anything more complicated. For example, we do have support for tags that are used to isolate scripts, such as `//go:build ignore`, via the [`"standaloneTags"`](https://github.com/golang/tools/blob/master/gopls/doc/settings.md#standalonetags-string) configuration, and that looks for files with a build constraint consisting of a single tag.

**Comment by meling** [comment]
@findleyr thanks for the pointer. Would it be possible to lift the restriction: *Gopls considers a file to be a standalone main file if and only if it has package name "main"* ?? Maybe then my use case would work. Or would it be preferable to support my use case via a separate configuration option? 

Either way I appreciate the work you are doing to improve the plugin and gopls. Thank you so much!

**Comment by arnevm123** [comment]
@findleyr I am trying out the new gopls and it seems to work perfectly.
The only thing that I'm wondering is if I'm using go to definition on a function that is both declared in a file with a Windows and Linux build flag it goes to the Linux implementation (which makes sense, I'm on Linux).
Is it meant to show both definitions (similar to go to implementation) or is this expected behaviour?

**Comment by findleyr** [comment]
@arnevm123 thanks for testing, and regarding your question this is expected behavior.  For each open file there is one "default" build, which is first and foremost the default build on your system, followed by the first match from [here](https://cs.opensource.google/go/x/tools/+/master:gopls/internal/lsp/cache/port.go;l=30;drc=a08f286e373b18182cab310ed596c0cc973c078f). This ordering is somewhat arbitrary, though first class ports are preferred. We do need a fixed order for determinism.

I'd prefer if we could return all definitions, but unfortunately for technical reasons this is really intractable. Note that a file without any build constraints matches all ports. We can't feasibly return every single definition across any port, because this would be much more expensive for little value in the common case. Note that the toolchain (Notably `go list` and the type checker) only operates on one build context at a time, so to answer questions about multiple build contexts we need to do additional work for each.  It is infeasible to do 50x the work (there are around 50 supported ports), so we'd have to try to be smarter by e.g. looking at ignored files in the same directory as the definition to see if there are other declarations of the same symbol. At that point our heuristic is liable to be flaky and/or buggy, and is tightly coupled to the go command (taking us even farther from supporting other build systems). Still, it may be worth pursuing at some point.

For now, I think we should keep it simple. We're going from gopls completely not working to mostly working, albeit with some surprises. Of course we'd greatly appreciate any alternative suggestions, or ways that we could make the current behavior more transparent.

**Comment by findleyr** [comment]
@meling 

> Gopls considers a file to be a standalone main file if and only if it has package name "main" ?? Maybe then my use case would work.

I think there's a fundamental difference between scripts, for which there may be many in the same directory, each consisting of a single file, and your use case, for which you want to have at most two versions of your package in the directory. In your use case, I assume you want gopls to resolve packages declarations contained in other unconstrained files in the same directory (for example a `common.go` file that is shared by both the `solution` and `!solution` builds).

**Comment by findleyr** [comment]
@meling I think the problem is that we could perhaps make `//go:build solution` work automatically by creating a new build with `-tags=solution` overlaid, but then users may be confused when opening a file tagged `//go:build (solution || experiment)` doesn't work automatically, and we don't want to get into the business of solving constraint equations.  The special handling for `//go:build ignore`, is justifiable because it's an extremely common pattern for scripts.

If enough users want this, we can reconsider. But for now, for the same reasons as https://github.com/golang/go/issues/29202#issuecomment-1889625558, let's keep it simple.

(and thanks again for the feedback!)

**Comment by findleyr** [comment]
gopls@v0.15.0 will be released soon (likely next week) containing the improvements discussed above. If you're interested to try it out, please install the latest prerelease, which is the release candidate:

```
go install golang.org/x/tools/gopls@v0.15.0-pre.4
```

(this differs from -pre.3 by just one minor CL)

With this release, I think working on multiple GOOS and GOARCH combinations is much better, though still not perfect. To follow up, I filed:

- #65755
- #65757
- #65758

This is in addition to @meling's:

- #65089

I think we're ready to close this issue in favor of the more specific follow-up items above. Please try out the release and let us know if you have feedback, either by commenting on or :+1:'ing one of the issues above, or filing a new issue describing the improvement you'd like to see (or bug you observe). Thanks!

**Comment by findleyr** [comment]
To clarify one point: the issue title "improve handling for build tags" could certainly be interpreted as also referring to better support for *custom* build tags, in addition to multiple GOOS/GOARCH. However, the issue description specifically mentions "completions, diagnostics, etc. for files tagged with any other OS", and I think we've achieved that in the v0.15.0 release.

I think better support for user-defined tags is covered by these two issues specifically:

- https://go.dev/issue/65089
- https://go.dev/issue/65757

Those issues both address ways that gopls could support working on multiple sets of user-defined tags. In #65089, it is proposed that gopls could automatically detect when there is one tag that separates files in a package. In #65757, it is proposed that users could explicitly configure multiple sets of tags.
